### DISCUSSION TITLE
having a specific DNS server for only one of my inbounds

### DISCUSSION THREAD

**Discussion by triumphantsoul** [XTLS/Xray-core#2501] [Q&A]
Hi everyone, I have a config file like the following, and I was wondering if there is a way to have only ONE of my inbounds go through the defined DNS server and not all of them!

{
  "log": {
    "loglevel": "info"
  },
"dns": {
    "servers": [
     "8.8.8.8"
    ]
  },
  "inbounds": [
{
      "tag": "VLESS TCP REALITY 1",
      "listen": "0.0.0.0",
      "port": 100,
      "protocol": "vless",
      "settings": {
        "clients": [],
        "decryption": "none"
      },
      "streamSettings": {
        "network": "tcp",
        "tcpSettings": {},
        "security": "reality",
        "realitySettings": {
          "show": false,
          "dest": "x.com:443",
          "xver": 0,
          "serverNames": [
            "x.com"
          ],
          "privateKey": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
          "shortIds": [
            "xxxxx"
          ]
        }
      },
      "sniffing": {
        "enabled": true,
        "destOverride": [
          "http",
          "tls"
        ]
      }
   },
{
      "tag": "VLESS TCP REALITY 2",
      "listen": "0.0.0.0",
      "port": 200,
      "protocol": "vless",
      "settings": {
        "clients": [],
        "decryption": "none"
      },
      "streamSettings": {
        "network": "tcp",
        "tcpSettings": {},
        "security": "reality",
        "realitySettings": {
          "show": false,
          "dest": "x.com:443",
          "xver": 0,
          "serverNames": [
            "x.com"
          ],
          "privateKey": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
          "shortIds": [
            "xxxxx"
          ]
        }
      },
      "sniffing": {
        "enabled": true,
        "destOverride": [
          "http",
          "tls"
        ]
      }
    },
    {
      "tag": "VLESS TCP REALITY 3",
      "listen": "0.0.0.0",
      "port": 300,
      "protocol": "vless",
      "settings": {
        "clients": [],
        "decryption": "none"
      },
      "streamSettings": {
        "network": "tcp",
        "tcpSettings": {},
        "security": "reality",
        "realitySettings": {
          "show": false,
          "dest": "x.com:443",
          "xver": 0,
          "serverNames": [
            "x.com"
          ],
          "privateKey": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
          "shortIds": [
            "xxxxx"
          ]
        }
      },
      "sniffing": {
        "enabled": true,
        "destOverride": [
          "http",
          "tls"
        ]
      }
    }
  ],
  "outbounds": [
    {
      "protocol": "freedom",
      "tag": "DIRECT"
    },
    {
      "protocol": "blackhole",
      "tag": "blackhole"
    }
  ],
  "routing": {
    "domainStrategy": "IPIfNonMatch",
    "rules": [
      {
        "type": "field",
        "outboundTag": "blackhole",
        "ip": [
          "geoip:private",
          "geoip:ir"
        ]
      },
      {
        "type": "field",
        "port": 53,
        "network": "tcp,udp",
        "outboundTag": "DNS-Internal"
      },
      {
        "type": "field",
        "outboundTag": "blackhole",
        "protocol": [
          "bittorrent"
        ]
      },
      {
        "outboundTag": "blackhole",
        "domain": [
          "regexp:.*\\.ir$",
          "ext:iran.dat:ir",
          "ext:iran.dat:other",
          "geosite:category-ir",
        ],
        "type": "field"
      }
    ]
  }
}

**Comment by us254** [comment]


**Reply by triumphantsoul** [reply]
thank you very much! I did exactly as you said but unfortunately it didn't work.

**Reply by us254** [reply]
To achieve the desired setup where only the inbound with the tag "VLESS TCP REALITY 1" uses the DNS server "8.8.8.8", while the other inbounds do not, you need to adjust both the `outbounds` and `routing` sections of your configuration. 

Here's a corrected version of the relevant parts of the JSON configuration:

### Modified Outbounds Section:

```json
"outbounds": [
  {
    "protocol": "freedom",
    "tag": "DIRECT"
  },
  {
    "protocol": "blackhole",
    "tag": "blackhole"
  },
  {
    "protocol": "dns",
    "tag": "DNS-Internal",
    "settings": {
      "address": "8.8.8.8",
      "port": 53,
      "network": "udp"
    }
  }
]
```

### Modified Routing Section:

```json
"routing": {
  "domainStrategy": "IPIfNonMatch",
  "rules": [
    {
      "type": "field",
      "inboundTag": ["VLESS TCP REALITY 1"],
      "network": "udp",
      "port": 53,
      "outboundTag": "DNS-Internal"
    },
    {
      "type": "field",
      "port": 53,
      "network": "udp",
      "inboundTag": ["VLESS TCP REALITY 2", "VLESS TCP REALITY 3"],
      "outboundTag": "DIRECT"
    },
    // ... other existing rules ...
  ]
}
```

### Explanation:

1. **Outbounds**: A new outbound with the tag "DNS-Internal" is added. This outbound is specifically configured to use the DNS server "8.8.8.8".

2. **Routing**:
   - The first rule directs UDP traffic on port 53 (DNS traffic) from the "VLESS TCP REALITY 1" inbound to the "DNS-Internal" outbound.
   - The second rule ensures that DNS traffic from "VLESS TCP REALITY 2" and "VLESS TCP REALITY 3" does not use the "DNS-Internal" outbound. Instead, it's handled by the default method (in this case, tagged as "DIRECT").

This setup ensures that only the inbound with the tag "VLESS TCP REALITY 1" will use the DNS server "8.8.8.8", while the other inbounds will use their standard DNS resolution methods.
